### ISSUE TITLE
Support for booting nodes in UEFI

### ISSUE THREAD

**Issue by geedo0** [CCI-MOC/hil#836]
First of all, I spoke with @naved001 and we agree that power_cycle() should not assume that the user wants the next boot device to be set to PXE. Instead, they should use the set_bootdev() routine prior to the boot. Further, with these specific options, ipmitool forces the boot mode into legacy BIOS mode which completely prevents any UEFI based boot chains with HIL.

.../obm/ipmi.py:72
``
    def power_cycle(self):
        self._ipmitool(['chassis', 'bootdev', 'pxe'])
        if self._ipmitool(['chassis', 'power', 'cycle']) == 0:
            return
``

Not sure how you guys want to support booting into UEFI, but here's the bug report from the Ironic team and some pointers on how they got it to work. Basically, the trick is to add "options=efiboot" to the ipmitool arguments.

https://bugs.launchpad.net/ironic/+bug/1611306

**Comment by zenhack** [comment]
We could add an (optional) field to the body of the request, so to force EFI boot, you'd do:

```
PUT /node/<node>/boot_device

{
    "bootdev": "disk",  (for example)
    "efi": true
}
```

**Comment by henn** [comment]
Now that we have set_bootdev(), I agree that it's no longer necessary to set a default boot device (to support that, the [deployment guide](https://github.com/CCI-MOC/hil/blob/master/docs/deployment.md) recommends having a particular boot order as well).

Regarding selecting UEFI, I agree with @zenhack - we can add it as an optional parameter to set_bootdev(). It might make sense to make the default a per-node tunable as well (maybe with out per-node metadata system). For example, there might be nodes in the future that don't support legacy booting that would fail to boot with our current setup. In that case, we'd want to assume `efi=true` by default.

**Comment by henn** [comment]
FYI - from [api.py/show_node()](https://github.com/CCI-MOC/hil/blob/af98ba26f79bc884c9189341210e21f05c6b1b3d/hil/api.py#L1139), it looks like any user can do a show_node() on a free node.

So I think storing it as a metadata item is viable.

**Comment by Izhmash** [comment]
Do we want this in the CLI too? Or is API/Client sufficient.

**Comment by zenhack** [comment]
Definitely need to expose it via the client library.

**Comment by Izhmash** [comment]
Current idea for the new structure: 
- A new boolean argument `efi` that determines Legacy vs. UEFI will be passed to `set_bootdev()`.
- `power_cycle` will no longer have the `_ipmitool()` PXE call on its first line.
- Nodes store Legacy/UEFI as a metadata item
- `node_set_bootdev` api call will take an optional `efi` argument.
    - If `efi` isn't set it will try to choose legacy/uefi from the node's boot metadata.  If all fails perhaps do nothing and send an error? Assuming the metadata is optional.

Questions:
- What happens if `set_bootdev()` isn't run before `power_cycle()`?

**Comment by zenhack** [comment]
I don't like the idea of hil directly acting on the metadata.

Re: what to do if set_bootdev isn't called, yeah, it will just default to whatever the BIOS is set to. The deployment docs suggest specifying a particular order, so I think that's OK.

**Comment by Izhmash** [comment]
Maybe instead then we do not have a default, and if someone calls `node_set_bootdev` without an `efi` option, it will request that the user choose one and display what the metadata option is set to?

**Comment by henn** [comment]
The issue here is that nodes can differ on what they support, so HIL needs a per-node way of representing that.

If we go with @Izhmash 's suggestion where we require the tenant to always specify an option, the tenant still needs a way to figure out what their node supports.

I don't see an easy/clean way of doing this other than using the metadata. Thoughts?

**Comment by SahilTikale** [comment]
HIL needs to define a namespace to keep per-node information that HIL core uses rather than provided by the OBM drivers.

**Comment by Izhmash** [comment]
I have the code written for UEFI booting, currently trying to figure out why ipmitool is failing to boot the Kumo PowerEdge M620's into UEFI.  Perhaps some machines support UEFI but don't support ipmitool's `efiboot`?

**Comment by geedo0** [comment]
It's important that the thing you're booting explicitly supports UEFI, otherwise the firmware will just failover and reboot. Have you successfully booted them with UEFI enabled manually yet?

**Comment by Izhmash** [comment]
Yeah, the M620's can boot UEFI if I set them manually in the BIOS setup.  I'm planning on testing this with another type of server soon to see if it's an issue with the machine or ipmitool.

**Comment by Izhmash** [comment]
So for the namespacing, we could have a metadata option such as `__bios` which can accept something such as `__legacy` or `__uefi`.  The API would check for this namespaced metadata in `node_set_bootev()`

**Comment by zenhack** [comment]
1. Probably better to make it a list; many machines support both mechanisms
2. There's no reason to apply the prefix to values; namespacing the keys should do the trick.
3. Nitpick: maybe pick something other than `__bios`, e.g. `__boot_firmware`; `__bios` is sometimes taken to imply the classic PC bios, whereas boot firmware could be anything that falls into that role (efi, the pcbios, openfirmware, something built around coreboot...)

**Comment by zenhack** [comment]
This probably belongs in obmd now...

**Comment by xuhang57** [comment]
we have lots of issues that are related to obmd,  or can be solved after we support obmd. Maybe we should group them together or leave them.

**Comment by zenhack** [comment]
obmd also has its own issue tracker: https://github.com/cci-moc/obmd/issues

I'm going to create a label for "needs obmd." Maybe we take those issues off of 0.3 in general, and then triage them together?

**Comment by naved001** [comment]
I remember @Izhmash telling me that he couldn't boot into uefi using ipmitool. He tested a dell m620 and a cisco node (the ones in kaizen/neu). So that's where we are stuck right now, if we can't do it using ipmitool then I don't know if there's any other way.

**Comment by zenhack** [comment]
Ok, yeah, obviously however we organize this architecturally we need to figure out the actual mechaism first.

I am going to go ahead and take this off the 0.3 milestone.
